File: JavaScript/FindTwoNonoverlappingSub-arraysEachWithTargetSum.js
const minSumOfLengths = function (arr, target) {
  const n = arr.length;

  let minLength = Infinity;
  let answer = Infinity;
  let total = 0;
  let best = Array(n).fill(Infinity);
  let j = 0;

  for (let i = 0; i < n; i++) {
    total += arr[i];

    while (total > target) {
      total -= arr[j];
      j++;
    }

    if (total === target) {
      const currentLength = i - j + 1;

      if (j > 0 && best[j - 1] !== Infinity) {
        answer = Math.min(answer, currentLength + best[j - 1]);
      }

      minLength = Math.min(minLength, currentLength);
    }

    best[i] = minLength;
  }

  return answer === Infinity ? -1 : answer;
};

console.log(minSumOfLengths([4, 3, 2, 6, 2, 3, 4], 6));

File: Python/FindTwoNon-overlappingSub-arraysEachWithTargetSum.py
class Solution:
    def minSumOfLengths(self, arr: list[int], target: int) -> int:

        n = len(arr)
        INF = n + 1
        best = [INF] * n
        sum = 0
        left = 0
        minLength = INF
        answer = INF

        for right in range(n):
            sum += arr[right]

            while sum > target:
                sum -= arr[left]
                left += 1

            if sum == target:
                currentLength = right - left + 1

                if left > 0 and best[left - 1] != INF:
                    answer = min(answer, currentLength + best[left - 1])

                minLength = min(minLength, currentLength)

            best[right] = minLength

        return -1 if answer == n + 1 else answer


solution = Solution()

solution.minSumOfLengths([3, 2, 2, 4, 3], 3)
